### PR TITLE
Get entry assembly name from host for GetProcessInfo2 command during suspension

### LINKED ISSUE #56808: Getting managed assembly entrypoint on diagnostic pipe GetProcessInfo2 command during suspension
### Description

When issuing a a GetProcessInfo2 command on the diagnostic pipe to a runtime instance that is suspended, it will fail to get the managed entrypoint assembly information and return an empty string. 

The expected behavior is that for any .NET process that does have a managed entrypoint, that this command will return the managed entrypoint assembly even while the process was suspended at startup.

Without this, it makes any diagnostic tool have an extremely difficult time determining if it should start observing a process based on its entrypoint name, especially in a multi-container environment where the process name is likely `dotnet` for all .NET processes and the process ID is 1 for many of them; the main differentiator will be the entrypoint name. Diagnostic tools would have start observing the process (start event pipe sessions, or maybe even attach a profiler, during suspension), have to "wait a little while" until the command does return a valid value for the entrypoint, and possibly cancel all of the diagnostic operations on the process if it was not one that the tool really wanted to observe.

Expectation is that the GetProcessInfo2 should be able to get the managed entrypoint assembly information during suspension. From a naive perspective, this shouldn't require running managed code but _just_ to read the assembly information from the metadata table.

### Regression?

No

### Other information

This is in reference to the following code: https://github.com/dotnet/runtime/blob/a33ea98572b93ad1b357961e2b2904605cdec9f6/src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.h#L1185

Behavior was introduced in the following PR: https://github.com/dotnet/runtime/pull/55379

cc @josalem

### PR DESCRIPTION
**Pull request by elinor-fung**
- Host runtime contract returns entry assembly name based on managed
  app being run, empty string for non-app (component) scenarios
- ProcessInfo2 command falls back to querying for entry assembly name
  from host

The added tests target the host and diagnostics/runtime separately, since the diagnostics tests don't use the product host. The host tests check that the host runtime contract returns the expected values. The diagnostics tests check that the ProcessInfo2 command can get the entry assembly name during suspension assuming the host (corerun in the tests) returns the expected value.

Resolves #56808

### CODE REVIEW

File: src/coreclr/hosts/corerun/corerun.cpp [hunk 1]
```diff
@@ -210,6 +211,35 @@ static void log_error_info(const char* line)
...
+        // Compute the entry assembly name based on the entry assembly full path
+        pal::string_t dir;
+        pal::string_t file;
+        pal::split_path_to_dir_filename(config->entry_assembly_fullpath, dir, file);
+        file = file.substr(0, file.rfind(W('.')));
+
+        pal::string_utf8_t file_utf8 = pal::convert_to_utf8(file.c_str());
+        size_t len = file_utf8.size() + 1;
+        if (value_buffer_size < len)
+            return len + 1;
```
Comment (vitek-karas) [inline src/coreclr/hosts/corerun/corerun.cpp:233]: The `+1` seems wrong here - we already added `+1` above to the length of the string. And also below we only require `len` length of the array.

**Review by vitek-karas — APPROVED** [review]
Other than the one comment this looks good.

File: src/coreclr/hosts/corerun/corerun.cpp (hunk 1 above)
Comment (elinor-fung) [reply src/coreclr/hosts/corerun/corerun.cpp:233]: Oops, I definitely failed at updating after switching `len` from not having the `+1` to having it.

File: src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.h [hunk 2]
```diff
@@ -1134,8 +1135,13 @@ ep_rt_entrypoint_assembly_name_get_utf8 (void)
 		}
 	}
 
-	// fallback to the empty string if we can't get assembly info, e.g., if the runtime is
```
Comment (davmason) [inline src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.h:1137]: Do we need both paths? Are there scenarios where the host won't set the assembly name?

Comment (elinor-fung) [reply src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.h:1137]: Our host will always set it - except for in non-app (component) scenarios, in which case there is no entry assembly so it is expected to be empty. However, if someone directly hosts coreclr, they don't need to set it - we discourage direct hosting and document it as obsolete, but there is still some functionality for which we don't have parity.

**Review by davmason — APPROVED** [review]
LGTM, thanks!

**Pull request merged**